File: example/1d-linear-convection/weno3/julia/03g/src/registry.jl
"""
统一注册系统 + 通用工厂
- ComponentRegistry: 组件注册表
- @register_component: 装饰器宏
- BaseFactory: 通用工厂接口
"""

module ComponentRegistry

# 注册表：Dict{category => Dict{name => constructor}}
const _REGISTRIES = Dict{String, Dict{String, Function}}()
const _VERBOSE = Ref(true)

# ---------------------- 注册核心逻辑 ----------------------
"""
    register(category::String, name::String, ctor::Function)

注册一个组件构造函数到指定类别。
如果已存在同名组件且不同，则发出警告。
"""
function register(category::String, name::String, ctor::Function)
    if !haskey(_REGISTRIES, category)
        _REGISTRIES[category] = Dict{String, Function}()
    end

    registry = _REGISTRIES[category]
    if haskey(registry, name)
        if registry[name] !== ctor && _VERBOSE[]
            @warn "覆盖注册: $category.$name"
        end
    end

    registry[name] = ctor
    if _VERBOSE[]
        println("✅ 已注册: $category.$name -> $(nameof(ctor))")
    end
end

"""
    create(category::String, name::String, args...; kwargs...)

从注册表创建组件实例。
"""
function create(category::String, name::String, args...; kwargs...)
    if !haskey(_REGISTRIES, category)
        error("❌ 未知类别: $category (可用: $(collect(keys(_REGISTRIES))))")
    end

    registry = _REGISTRIES[category]
    lname = lowercase(name)
    if !haskey(registry, lname)
        available = sort(collect(keys(registry)))
        error("❌ 未找到: $category.$name (可用: $available)")
    end

    return registry[lname](args...; kwargs...)
end

"""
    list_all()

返回所有已注册组件（按类别）。
"""
function list_all()
    return Dict(cat => sort(collect(keys(reg))) for (cat, reg) in _REGISTRIES)
end

"""
    set_verbose(flag::Bool)

开启/关闭注册提示。
"""
function set_verbose(flag::Bool)
    _VERBOSE[] = flag
end

end  # module ComponentRegistry


# ---------------------- 装饰器宏：@register_component ----------------------
"""
@register_component(category, [name])

用法：
  @register_component("boundary", "periodic")
  struct PeriodicBoundary ...

若省略 name，则使用类型名的小写形式。
"""
macro register_component(category::String, name_expr)
    error("@register_component 需在类型定义前使用，且必须在模块顶层")
end

macro register_component(category::String)
    error("@register_component(category, name) 需指定 name 或在类型后使用")
end

# 重载：@register_component("category", "name") struct X ... end
macro register_component(category::String, name::String, ex)
    if !Meta.isexpr(ex, :struct)
        error("@register_component 必须作用于 struct 定义")
    end

    struct_name = ex.args[2]
    if Meta.isexpr(struct_name, :curly)
        struct_name = struct_name.args[1]
    end

    # 插入注册调用（在模块顶层）
    quote
        $(esc(ex))
        $(ComponentRegistry).register($(category), $(name), $(esc(struct_name)))
    end
end

# 重载：@register_component("category") struct X ... end → name = lowercase(nameof(X))
macro register_component(category::String, ex)
    if !Meta.isexpr(ex, :struct)
        error("@register_component 必须作用于 struct 定义")
    end

    struct_name = ex.args[2]
    if Meta.isexpr(struct_name, :curly)
        struct_name = struct_name.args[1]
    end

    name_str = string(struct_name) |> lowercase

    quote
        $(esc(ex))
        $(ComponentRegistry).register($(category), $(name_str), $(esc(struct_name)))
    end
end


# ---------------------- 通用工厂 ----------------------
module BaseFactory

using ..ComponentRegistry

"""
    create_component(category::String, name::String, args...; kwargs...)

通用工厂接口，与 Python BaseFactory.create_component 行为一致。
"""
function create_component(category::String, name::String, args...; kwargs...)
    return ComponentRegistry.create(category, name, args...; kwargs...)
end

"""
    get_available_components(category::String)

列出某类别下所有可用组件。
"""
function get_available_components(category::String)
    all = ComponentRegistry.list_all()
    return get(all, category, String[])
end

end  # module BaseFactory


# 导出接口
export ComponentRegistry, BaseFactory, @register_component

File: example/1d-linear-convection/weno3/python-v1/01h/core/registry.py
# core/registry.py
"""
统一注册系统 + 通用工厂
- ComponentRegistry: 组件注册表
- register_component: 装饰器
- BaseFactory: 通用工厂类
"""

from typing import Dict, Type, Any


# ==================== 1. 注册表核心 ====================
class ComponentRegistry:
    """组件注册表"""
    _registries: Dict[str, Dict[str, Type]] = {}
    _verbose = True

    @classmethod
    def set_verbose(cls, verbose: bool):
        cls._verbose = verbose

    @classmethod
    def register(cls, category: str, name: str, component_class: Type):
        #print(f"ComponentRegistry register cls={cls}")
        #print(f"ComponentRegistry register name={name}")
        #print(f"ComponentRegistry register component_class={component_class}")
        if category not in cls._registries:
            cls._registries[category] = {}
        if name in cls._registries[category]:
            if cls._registries[category][name] != component_class and cls._verbose:
                print(f"⚠️  覆盖注册: {category}.{name}")
        cls._registries[category][name] = component_class
        if cls._verbose:
            print(f"✅ 已注册: {category}.{name} -> {component_class.__name__}")

    @classmethod
    def get(cls, category: str, name: str) -> Type:
        if category not in cls._registries:
            raise ValueError(f"❌ 未知类别: {category} (可用: {list(cls._registries.keys())})")
        if name not in cls._registries[category]:
            raise ValueError(f"❌ 未找到: {category}.{name} (可用: {list(cls._registries[category].keys())})")
        return cls._registries[category][name]

    @classmethod
    def create(cls, category: str, name: str, *args, **kwargs) -> Any:
        component_class = cls.get(category, name)
        return component_class(*args, **kwargs)

    @classmethod
    def list_all(cls):
        return {cat: list(comps.keys()) for cat, comps in cls._registries.items()}


# ==================== 2. 装饰器 ====================
def register_component(category: str, name: str = None):
    """简化注册的装饰器"""
    def decorator(cls):
        component_name = name or cls.__name__.lower()
        #print(f"register_component  decorator name={name}")
        ComponentRegistry.register(category, component_name, cls)
        return cls
    return decorator


# ==================== 3. 通用工厂 ====================
class BaseFactory:
    """通用工厂基类"""
    @classmethod
    def create_component(cls, category: str, name: str, *args, **kwargs) -> Any:
        name_lower = name.lower()
        try:
            return ComponentRegistry.create(category, name_lower, *args, **kwargs)
        except ValueError as e:
            available = ComponentRegistry.list_all().get(category, [])
            if available:
                error_msg = f"不支持的 {category} 类型 '{name}'。可用类型：{available}"
            else:
                error_msg = f"不支持的 {category} 类型 '{name}'（无已注册组件）"
            raise ValueError(error_msg) from e

    @classmethod
    def get_available_components(cls, category: str) -> list:
        return ComponentRegistry.list_all().get(category, [])